consensus/src/units/mod.rs

```diff
@@ -218,15 +218,73 @@ impl<H: Hasher> Unit<H> {
     }
 }
 
+#[cfg(test)]
+pub use tests::{
+    add_units, create_units, creator_set, preunit_to_unchecked_signed_unit, preunit_to_unit,
+};
+
 #[cfg(test)]
 mod tests {
     use crate::{
-        units::{ControlHash, FullUnit, PreUnit},
-        Hasher, NodeIndex,
+        creation::Creator as GenericCreator,
+        units::{
+            ControlHash, FullUnit as GenericFullUnit, PreUnit as GenericPreUnit,
+            UncheckedSignedUnit as GenericUncheckedSignedUnit, Unit as GenericUnit,
+        },
+        Hasher, NodeCount, NodeIndex, Round, SessionId, Signed,
     };
-    use aleph_bft_mock::Hasher64;
+    use aleph_bft_mock::{Data, Hasher64, Keychain, Signature};
     use codec::{Decode, Encode};
 
+    type Creator = GenericCreator<Hasher64>;
+    type PreUnit = GenericPreUnit<Hasher64>;
+    type Unit = GenericUnit<Hasher64>;
+    type FullUnit = GenericFullUnit<Hasher64, Data>;
+    type UncheckedSignedUnit = GenericUncheckedSignedUnit<Hasher64, Data, Signature>;
+
+    pub fn creator_set(n_members: NodeCount) -> Vec<Creator> {
+        let mut result = Vec::new();
+        for i in 0..n_members.0 {
+            result.push(Creator::new(NodeIndex(i), n_members));
+        }
+        result
+    }
+
+    pub fn create_units<'a, C: Iterator<Item = &'a Creator>>(
+        creators: C,
+        round: Round,
+    ) -> Vec<(PreUnit, Vec<<Hasher64 as Hasher>::Hash>)> {
+        let mut result = Vec::new();
+        for creator in creators {
+            result.push(
+                creator
+                    .create_unit(round)
+                    .expect("Creation should succeed."),
+            );
+        }
+        result
+    }
+
+    pub fn preunit_to_unit(preunit: PreUnit, session_id: SessionId) -> Unit {
+        FullUnit::new(preunit, 0, session_id).unit()
+    }
+
+    pub fn add_units(creator: &mut Creator, units: &[Unit]) {
```

Comment:
could use method chaining here instead of fugly interior mutbility:
```
impl for Creator {
  add_units(creator: &mut Creator, units: &[Unit]) -> Self {
    for unit in units {
      creator.add_unit(unit);
    }
    self.to_owned()
  }
}
```
and then:
 
```creator.add_units(units);```

Reply:
I added this method for testing. I'm just not sure what for is chaining here. There is no case where this would be needed.